python/src/iceberg/io/s3.py

```diff
@@ -0,0 +1,185 @@
+        self._s3_client = kwargs.get("s3_client")
+        return self
+
+    def __len__(self):
+        s3 = self._create_s3_session() if not self._s3_client else self._s3_client
+        s3_url = urlparse(self.location)
+        resource = s3.head_object(Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/"))
+        return resource["ContentLength"]
+
+    def _create_s3_session(self):
+        return boto3.Session().client("s3")
+
+    def exists(self, s3_client=None):
+        s3 = self._create_s3_session() if not s3_client else s3_client
+        s3_url = urlparse(self.location)
+        try:
+            s3.head_object(Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/"))
+        except ClientError as e:
+            if "Not Found" in str(e):
+                return False
+            else:
+                raise
+
+        return True
+
+    def to_input_file(self):
+        return S3InputFile(location=self._location)
+
+    def write(self):
+        s3 = self._create_s3_session() if not self._s3_client else self._s3_client
+        s3_url = urlparse(self.location)
+        return s3.upload_fileobj(
+            Fileobj=self._file_obj, Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/")
+        )
+
+    def __enter__(self):
+        """Enter context for S3InputFile
+
+        This returns a botocore.response.StreamingBody instance for
+        the location. If an `s3_client` is not provided, a boto3 session
+        using the environment AWS credentials will be created.
+        """
+        super().__enter__()
+        if not self._file_obj:
+            raise RuntimeError(
+                "file_obj was not set. When using this context, use a call to set"
+                'a file object: `with S3OutputFile(file_obj="s3://foo/bar.json") as f:`'
+            )
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """A NOOP since boto3 automatically handles closing the session"""
+        super().__exit__(exc_type, exc_value, exc_traceback)
+        del self._file_obj
+        return
+
+
+class S3FileIO(FileIO):
+    def __init__(self, s3_client=None):
+        self._s3_client = boto3.Session().client("s3") if not s3_client else s3_client
+
+    def new_input(self, location: str):
+        super().new_input(location=location)
+        return S3InputFile(location=location)
+
+    def new_output(self, location: str):
+        super().new_output(location=location)
+        return S3OutputFile(location=location)
+
+    def delete(self, location: Union[str, S3InputFile, S3OutputFile]):
+        delete_location = (
+            location.location
+            if isinstance(location, (S3InputFile, S3OutputFile))
+            else location
+        )
+        super().delete(location=delete_location)
+        s3_url = urlparse(delete_location)
+        s3 = self._s3_client
+        return s3.delete_object(Bucket=s3_url.netloc, Key=s3_url.path.lstrip("/"))
```

Comment:
What does `delete_object` return?

Reply:
It returns the aws response (just headers) in the form of a python dictionary. Here's a sample response from the [aws docs](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObject.html):
```
           HTTP/1.1 204 NoContent
           x-amz-id-2: LriYPLdmOdAiIfgSm/F1YsViT1LW94/xUQxMsF7xiEb1a0wiIOIxl+zbwZ163pt7
           x-amz-request-id: 0A49CE4060975EAC
           Date: Wed, 12 Oct 2009 17:50:00 GMT
           Content-Length: 0
           Connection: close
           Server: AmazonS3
```